apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,565 @@
+      span.textContent = matchString;
+      node.appendChild(span);
+    }
+
+    var endString = text.substring(endIndex);
+    if (endString) {
+      node.appendChild(document.createTextNode(endString));
+    }
+  } else {
+    node.textContent = text;
+  }
+}
+
+/**
+ * Given a node, if it or an ancestor has the CSS className, return the node
+ * with that className. Returns null if not a match.
+ * @param  {Node} node
+ * @return {Node}
+ */
+function findAncestorWithClass(node, className) {
+  if (node.classList && node.classList.contains(className)) {
+    return node;
+  }
+
+  while ((node = node.parentNode)) {
+    if (node.classList && node.classList.contains(className)) {
+      return node;
+    }
+  }
+
+  return null;
+}
+
+/**
+ * Given a node, if it or an ancestor is an autocomplete item in the
+ * autocomplete list, return the autcomplete item node. Returns null if not a
+ * match.
+ * @param  {Node} node
+ * @return {Node}
+ */
+function getAutocompleteItem(node) {
+  return findAncestorWithClass(node, 'cmp-autocomplete-item');
+}
+
+/**
+ * Autocomplete custom element that contains a list of elements that show the
+ * autocomplete matches.
+ *
+ * This custom element relies on a CSS class-based API to bind to nodes outside
+ * of this autocomplete. The CSS/HTML structure expected by this autocomplete:
+ *
+ * - An element with the CSS class of cmp-autocomplete-origin" that is the the
+ * master scroll area for the view. It should be an ancestor node to the
+ * autocomplete and also contain the 'cmp-autocomplete-input-list' element that
+ * has all the input elements. This element is used as the "origin" for the
+ * style.top placement of the autocomplete list.
+ *
+ * - An element that holds all the input fields, with the CSS class of
+ * "cmp-autocomplete-input-list". That inputArea will be translateY'd to make
+ * sure the current input focus is visible at the top of the view.
+ *
+ * The autocomplete custom element also expects a "data-source" HTML attribute
+ * to be set on it, which indicates the module ID to use as the data source for
+ * the matches.
+ *
+ * The owner of this custom element instance (typically a view card), should
+ * override the getExistingEntries() method on the instance to return the
+ * current entries, to avoid dupes showing up in subsequent autocomplete
+ * matches.
+ */
+return [
+  require('../base')(),
+  {
+    createdCallback: function() {
+      //TODO: master hack only. In email-model-instance branch, this is not
+      //needed. In the branch, there is a separate 'base' that elements can
+      //mix in that is not assumed to be a card. For now, this is an equivalent
+      //result.
+      this.classList.remove('card');
+
+      this.lastTranslateY = 0;
+      this.inputAreaTranslated = Promise.resolve();
+
+      this.runQuery = this.runQuery.bind(this);
+      this.onFocus = this.onFocus.bind(this);
+      this.onDocumentEvent = this.onDocumentEvent.bind(this);
```

Comment:
I was very close to create a helper that accepted a list of method names and it would do this work, but I did not want to disentangle it from this version of base.js from the one on the convoy branch.

On first thought, I'm lukewarm on auto-binding for all methods, seems excessive. I suppose we can talk more about it in convoy.

I was hoping via other APIs to avoid the need for these kinds of explicit .bind() calling. evt now takes object listeners on the convoy branch, so this binding is no longer necessary for those cases, and the data-event binding takes care of it in its cases. Dojo's event module also could handle DOM addEventListener registration in a similar way to the evt object listening, so that might be another option to address this particular case, where it is about DOM event listening vs the non-DOM evt listening.

This case is a bit weird because it is reaching outside of itself to listen to things.